src/scimlfunctions.jl

```diff
@@ -289,10 +290,16 @@ the usage of `f`. These include:
   based on the sparsity pattern. Defaults to `nothing`, which means a color vector will be
   internally computed on demand when required. The cost of this operation is highly dependent
   on the sparsity pattern.
-- `nlprob`: a `NonlinearProblem` that solves `f(u, t, p) = u_tmp`
-  where the nonlinear parameters are the tuple `(t, u_tmp, p)`.
-  This will be used as the nonlinear problem inside an implicit solver by specifying `u, u_tmp` and `t`
-  such that solving this function produces a solution to the implicit step of your solver.
+- `nlprob`: some `AbstractNonlinearProblem` to define custom nonlinear problems to be used for
+  implicit time discretizations. This allows to use extra structure of the ODE function (e.g.
+  multi-level structure). The rationale here is essentially that implicit ODE integration algorithms
+  need do solve the a nonlinear problems, usually of the form `dt⋅f(γ⋅z+inner_tmp,p,t) + outer_tmp = z`,
+  during time integration, where users want to use extra structure of f during the nonlinear solve.
+  Here `z` is the stage solution vector, `p` is the parameter vector of the ODE problem, `t` is
```

Comment:
how are dt, inner_tmp, outer_tmp, and z defined? in `p`? Where in `p` is it assumed? Or does it match some interface?

Aren't you missing a function that updates the prob based on those values?

Reply:
Good point. Tried to resolve this in the latest commit.

Reply:
I don't see how that addresses it. So okay, how do I update `inner_tmp`?

Reply:
I do not see how that logic changes in OrdinaryDiffEqNonlinearSolve.jl. From my current understanding the specific ode integrator at hand passes some struct with e.g. `inner_tmp` into the `nlprob` (e.g. via remake?) to update these during setup and remake.

Reply:
Okay that struct is currently just a tuple. I specifically refer to https://github.com/SciML/OrdinaryDiffEq.jl/blob/1d05b16b9bc52c31f5ec27e347b77d9e5306bee1/lib/OrdinaryDiffEqNonlinearSolve/src/newton.jl#L46-L47 where the tmp variable comes from the nonlinear solver struct in https://github.com/SciML/OrdinaryDiffEq.jl/blob/1d05b16b9bc52c31f5ec27e347b77d9e5306bee1/lib/OrdinaryDiffEqNonlinearSolve/src/type.jl#L216-L225

Reply:
Okay is that the interface? If you don't build the nlprob such that it is true will its integrations break?

Reply:
> Okay is that the interface?

From my understanding. Can you confirm or clarify if I got some detail wrong @oscardssmith ?

> If you don't build the nlprob such that it is true will its integrations break?

I am pretty sure that it will break the integration. Especially if we stick to the tuple instead of a struct and change the solver to something which requires a slightly different tuple, which is compatible by accident.

Reply:
Tuple can be quite limiting though. Why would we not have a functional interface rather than specifying a data type?

Reply:
We probably should have a better interface (a `NamedTuple` at least would be a step up)

Reply:
I'm not sure a functional interface buys us much here since the only things the user is expected to do is get the elements out...

Reply:
MTK cannot easily generate a NamedTuple. An interface via a function would make it easy to generate via SII

Reply:
> [...] functional interface [...]

Functional interface as in _non-defective interface_ or as in _function programming_?

Regarding the tuples, I think we are on the same page then. I see two main advantage of having structs to control the dispatches for the initialization of different nonlinear solver classes (over having tuples and hard-coded if-else blocks on types):
1. Being extensible for custom solvers for specific problem classes
2. Being less error prone to changes, e.g. when there might be some decision in the future to add another possibility to the values of `method`.